### ISSUE TITLE
Problem running ib_write_bw_gdr test on AKS+NDv4

### ISSUE THREAD

**Issue by garvct** [Azure/azurehpc-health-checks#46]
I was not able to achieve the expected IB network BW of 185 Gbps on an AKS with NDv4 (40GB A100) running an Ubuntu 22.04 hosting OS using the check_ib_write_bw_gdr test, I was only able to achieve about ~30 Gbps. I was able to achieve ~194 Gbps  with a couple of modifications.

1.  I believe the current  numactl pinning used is not correct, its currently pinned based on the location of GPU's and not pinned based on the location of the IB devices. e.g. for NDv4, I changed the pinning array from (1 1 0 0 3 3 2 2), to (3 3 2 0 2 1 0 1)
2. The current test runs a sort of loop-back ib_write_bw test on each of the IB devices in parallel, I changed this to do 4 serial ib_write_bw tests on 4 IB device pairs (0,4), (1,5), (2,6), (3,7)
3. Removed the --use-cuda ib_write_bw argument.
4. I added a re-try logic to allow a couple of re-try's before determining the test has failed.
5. The modified code is here https://github.com/Azure/azurehpc/blob/master/experimental/aks_ndv4/azure_ib_write_bw_gdr.nhc 

Questions
1. Would be interested in your feedback on the above mods?
2. I suspect we may not need a gdr and non_gdr verion of ib_write_bw
3. Running the ib_write_bw benchmark in pairs of IB devices hit more H/W and be a better test than running it in a loop-back mode on all 8 IB devices?

**Comment by rafsalas19** [comment]
Question 1, Let me investigate this and I'll get back to you.

To address question 2, I was considering this as well, but this change wouldn't be high priority at the moment. I can probably just add/modify this on an upcoming PR.

To address question 3. I agree on the hardware coverage. However, given the NDv5 NDR topology the traffic actually travels up to the switch layers and then back (in the case of a HCA pair wise test). Since we wanted to keep NHC tests within a node the choice was made to make ib_write perform a loop back instead of performing IB write pair tests. In the interest of standardizing the tests, the NDv4 test is similar to NDv5. 

Give me a bit of time and I'll give you better feed back

**Comment by rafsalas19** [comment]
1. I have check the numa domains using lstopo and it shows this is the correct topo for ndv4: "(1 1 0 0 3 3 2 2)". Am I missing something?
2. Changes to GDR:
  - I have added a bidirectional flag to ibwrite gdr. This has increased the returning throughput
  - Installation now just installs perf-test w/ CUDA, which works for both gdr/non-gdr
 3. As stated before we are not going to test ib_write in pairs due to IB communications travelling up to the IB2 layer for NDR topology. However, the "check_nccl_allreduce_ib_loopback" test is ran for ndv4. This tests cover HCA to HCA communications. 

4. I snuck some changes into this PR: https://github.com/Azure/azurehpc-health-checks/pull/50, I will address retry logic and maybe consolidating the ib_write test in a separate PR.

**Comment by garvct** [comment]
In regards to 1), the issue is in AKS (ubuntu 22.04) the lstopo is different than the ubuntu-hpc marketplace image. (Will need to investigate why).
I assume  in 2) you meant w/o(without) CUDA ?
Will try the updated version of ib_write_bw when available. Thanks.

**Comment by rafsalas19** [comment]
Regarding 1, I wasn't aware of the discrepancy. lstopo should show correct topo. Do you know if this is image specific, region specific? We can take this off-line, but it might be helpful if I can repro it outside of aks.

**Comment by rafsalas19** [comment]
I have modified the custom tests to move away from the hard coded IB naming. However, for SKUs with different naming conventions the conf files should be modified to change the default IB port checks(``` * || check_hw_ib 200 mlx5_ib1:1```). The naming convention on the AKS image should be the same as AzHPC images. If it is not then we should use custom conf files for AKS images. I know you mentioned using the naming script from AzHPC, if there are any issues with this please let me know.

**Comment by rafsalas19** [comment]
This has been fully addressed in v0.2.8.
